File: code-sample-java/spring-boot/code-sample-spring-security/src/main/java/codesample/springboot/security/UserService.java
package codesample.springboot.security;

import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@Service
public class UserService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    public UserService(UserRepository userRepository, PasswordEncoder passwordEncoder) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
    }

    /* By the time this check is executed, the authentication object has already been constructed
    (including all the checks such as is user enabled / are credentials expired, etc)
    * thus no password check or anything alike is required*/
    public User getCurrentAuthenticatedUser() {
        Object principal = SecurityContextHolder.getContext().getAuthentication().getPrincipal();
        UserDetails userDetails = (UserDetails) principal;
        return userRepository.findByUsername(userDetails.getUsername()).orElseThrow(UserNotFoundException::new);
    }

    public User createNewAndSave(String username, String password) {
        String encodedPassword = passwordEncoder.encode(password);
        User user = User.createNewEnabled(username, encodedPassword);
        return userRepository.save(user);
    }
}

File: code-sample-kotlin/code-sample-kotlin-spring-security/src/main/kotlin/codesample/kotlin/service/UserService.kt
package codesample.kotlin.service

import codesample.kotlin.entity.User
import codesample.kotlin.repository.UserRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service

@Service
class UserService (@Autowired private val userRepository: UserRepository,
                   @Autowired private val passwordEncoder: PasswordEncoder){

    fun getCurrentAuthenticatedUser() : User {
        /* This principle will be obtained from DbUserDetailsService because we implement authenticationProvider
        * and set it as UserDetailsService for auth provider */
        val principal = SecurityContextHolder.getContext().authentication.principal
        return principal as User
    }

    fun createNewAndSave(username: String, password: String, secondName: String): User {
        val encodedPassword = passwordEncoder.encode(password)
        val user = User(username = username,
                password = encodedPassword,
                secondName = secondName)
        return userRepository.save(user)
    }
}